os/kernel/binary_manager/binary_manager.c

```diff
@@ -0,0 +1,380 @@
+	strncpy(BIN_VER(bin_idx), header_data[latest_idx].bin_ver, BIN_VER_MAX);
+	strncpy(BIN_KERNEL_VER(bin_idx), header_data[latest_idx].kernel_ver, KERNEL_VER_MAX);
+	strncpy(BIN_NAME(bin_idx), header_data[latest_idx].bin_name, BIN_NAME_MAX);
+
+	bmvdbg("BIN TABLE[%d] %d %d %s %s %s\n", bin_idx, BIN_SIZE(bin_idx), BIN_RAMSIZE(bin_idx), BIN_VER(bin_idx), BIN_KERNEL_VER(bin_idx), BIN_NAME(bin_idx));
+
+	return OK;
+}
+
+/* Load binary with index in binary table */
+int binary_manager_load_binary(int bin_idx)
+{
+	int ret;
+	pid_t pid;
+	char devname[BINMGR_DEVNAME_LEN];
+
+	if (bin_idx < 0) {
+		bmdbg("Invalid bin idx %d\n", bin_idx);
+		return ERROR;
+	}
+
+	/* Load binary info */
+	ret = binary_manager_load_bininfo(bin_idx);
+	if (ret < 0) {
+		bmdbg("Failed to get loadable binary %d\n", bin_idx);
+		return ERROR;
+	}
+
+	/* Load binary */
+	snprintf(devname, BINMGR_DEVNAME_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, BIN_USEIDX(bin_idx)));
+	bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, BIN_SIZE(bin_idx), BIN_OFFSET(bin_idx));
+	pid = load_binary(devname, BIN_SIZE(bin_idx), BIN_OFFSET(bin_idx), BIN_RAMSIZE(bin_idx));
+	if (pid <= 0) {
+		bmdbg("Load '%s' fail, errno %d\n", BIN_NAME(bin_idx), errno);
+		return ERROR;
+	}
+
+	BIN_ID(bin_idx) = pid;
+	bmvdbg("Load '%s' success! pid = %d\n", BIN_NAME(bin_idx), BIN_ID(bin_idx));
+
+	return OK;
+}
+
+/* Create recovery thread to handle fault with pid */
+int binary_manager_reload_binary(binmgr_request_t *request_msg)
+{
+	return OK;
+}
+
+/* Get index in binary table with binary id */
+int binary_manager_get_index_with_binid(int bin_id)
+{
+	int bin_idx;
+
+	if (bin_id <= 0) {
+		bmdbg("Invalid bin_id %d\n", bin_id);
+		return ERROR;
+	}
+
+	for (bin_idx = 1; bin_idx <= g_bin_count; bin_idx++) {
+		if (BIN_ID(bin_idx) == bin_id) {
+			bmdbg("FIND binary %d idx = %d\n", bin_id, bin_idx);
+			return bin_idx;
+		}
+	}
+
+	return ERROR;
+}
+
+/* Get binary info with binary name */
+int binary_manager_get_info_with_name(char *bin_name)
+{
+	return OK;
+}
+
+/* Get info of all registered binaries */
+int binary_manager_get_info_all(void)
+{
+	return OK;
+}
+
+/****************************************************************************
+ * Main Function
+ ****************************************************************************/
+int binary_manager(int argc, char *argv[])
+{
+	int ret;
+	int nbytes;
+	int bin_idx;
+	int load_cnt;
+	binmgr_request_t request_msg;
+	mqd_t binmgr_mq_fd;
+
+	struct mq_attr attr;
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_request_t);
+	attr.mq_flags = 0;
+
+	bmvdbg("Binary Manager STARTED\n");
+
+	/* Create binary manager message queue */
+	binmgr_mq_fd = mq_open(BINMGR_REQUEST_MQ, O_RDONLY | O_CREAT, 0666, &attr);
+	if (binmgr_mq_fd < 0) {		
+		bmdbg("Failed to open message queue\n");
+		return 0;
+	}
+
+	/* Load all user binaries */
+	load_cnt = 0;
+	for (bin_idx = 1; bin_idx <= g_bin_count; bin_idx++) {
```

Comment:
If loaded binary makes fault, binary manager cannot handle the fault immediately.
Please make thread or change the logic.